### ISSUE TITLE
`unnecessary_lazy_evaluations` does not take into account utility of laziness to avoid addition/subtraction over/underflowing

### ISSUE THREAD

**Issue by rben01** [rust-lang/rust-clippy#11961]
### Summary

Sample code:

```rust
let x = 0_usize;
let y = (x > 0).then(|| x - 1);
```

This triggers `clippy::unnecessary_lazy_evaluations`, and the suggestion is to replace with `(x > 0).then_some(x - 1)`. However, this will lead to underflow, which panics in debug builds. Leaving the closure is the correct behavior.

### Lint Name

unnecessary_lazy_evaluations

### Reproducer

I tried this code:

```rust
let x = 0_usize;
let y = (x > 0).then(|| x - 1);
```

I saw this happen:

```
unnecessary closure used with `bool::then`
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
use `then_some(..)` instead: `then_some(x - 1)`
```

I expected to see this happen:

In this case, lazily evaluating `x-1` is crucial because if it underflows, it will panic in debug builds.

### Version

```text
rustc 1.76.0-nightly (9a66e4471 2023-11-19)
binary: rustc
commit-hash: 9a66e4471f71283fd54d80ef8147630d34756332
commit-date: 2023-11-19
host: aarch64-apple-darwin
release: 1.76.0-nightly
LLVM version: 17.0.5
```

**Comment by y21** [comment]
See #11931 and #11930. This should be fixed on more recent nightly builds.

**Comment by rben01** [comment]
Thanks.
